tools-in-rust-repo.md

```diff
@@ -0,0 +1,139 @@
+# Maintaining tools in the Rust repository
+
+This document contains information for tool maintainers and Rust contributors
+who need to work with tools that are part of the Rust distribution and
+repository. For information on requirements for such tools see the
+[stability docs](https://github.com/nrc/dev-tools-team/blob/master/stability/README.md).
+
+When adding tools or making significant changes, please be sure to cc @nrc and
+@aclexcrichton on PRs. Adding tools to the Rust repo requires approval from the
+dev-tools team and sign-off from the core team.
+
+
+## Repository layout
+
+Tools should live in their own repository which should be in an official
+organisation (e.g., [rust-lang-nursery](https://github.com/rust-lang-nursery)).
+Tools should be included into the Rust repository as a Git submodule in the
+[src/tools](https://github.com/rust-lang/rust/tree/master/src/tools) directory.
+
+
+## Constraints on tools
+
+It is essential that any commit used as a submodule in the Rust repo is
+preserved forever (i.e., is *never* deleted or changed). This ensures that we
+will continue to be able to build any historic release of Rust (which is
+important for bisecting regressions, etc). In order to do this, any commit
+pulled into the Rust repo as a submodule must be from a Rust organisation repo,
+not an author's. If the commit is only on a branch, that branch cannot be
+deleted. Whether the commit is on a branch or master, it cannot be rebased
+(which essentially deletes the commit and creates a new one).
+
+Tools used in the Rust repo cannot use Cargo workspaces. This is because the
+tool itself will be treated as a workspace crate and workspaces cannot be
+nested.
+
+
+## Making breaking changes
+
+If making a change to a 'core' part of Rust causes breakage in any of the tools,
+then there is a bit of a dance to perform to land the Rust PR.
+
+Communication is key here! If you're a compiler dev, communicate with tool devs
+as soon as possible (and vice versa).
+
+At the end of the day, the tool's maintainers are responsible for ensuring that
+their tool continues to build and pass tests in the Rust repo. For non-trivial
+fixes, tool devs should be happy to help fix breaking changes.
+
+How to land a breaking change:
```

Comment:
I guess? At some point I'm not personally responsible for the tools at hand, so I'll let y'all call the shots...

I really am worried about the contributor experience, though. I'd dread wanting to make a change to rustc only realizing it breaks 3 tools so I have to go and update 3 submodules, push to new branches, coordinate with maintainers, etc. If this starts happening a lot it seems like it can become a real drag on iterating in the compiler?